File: r/tests/testthat/test-hist.R
test_that("valid 'from_date' and 'interval'", {

  skip("long-running test")
  
  fields <- yfhist::data_intervals[["field"]]
  
  count <- 0
  result_ls <- list()
  errors_ls <- list()
  
  for (field in fields) {
    
    lookback <- yfhist::data_intervals[["lookback"]][yfhist::data_intervals[["field"]] == field]
    
    if (is.na(lookback)) {
      lookback <- NULL
    }
    
    to_date <- as.POSIXct(Sys.time(), tz = "UTC")
    
    if (is.null(lookback)) {
      from_date <- "2007-01-01"
    } else {
      if (field == "1m") {
        from_date <- as.POSIXct(to_date - 8 * 24 * 3600, tz = "UTC")
      } else {
        from_date <- as.POSIXct(to_date - (lookback - 1) * 24 * 3600, tz = "UTC")
      }
    }
    
    for (symbols in test_symbols) {
      
      response <- tryCatch({
        
        data <- suppressWarnings(get_data(symbols, from_date = from_date,
                                          to_date = to_date, interval = field))
        
        for (col in test_cols) {
          response <- get_col(data, col)
        }
        
        # if (is.null(response)) {
        #   response <- "success"
        # } else {
        response
        # }
        
      }, error = function(e) {
        NULL
      })
      
      if (is.null(response)) {
        
        error <- data.frame(
          symbols = paste(symbols, collapse = ","),
          field = field
        )
        
        errors_ls <- append(errors_ls, list(error))
        
      }
      
      count <- count + 1
      
      if (count %% 5 == 0) {
        
        message("pause one second after five requests")
        Sys.sleep(1)
        
      }
      
    }
    
  }
  
  if (length(errors_ls) > 0) {
    
    result <- do.call(rbind, errors_ls)
    result_ls <- append(result_ls, list(result))
    
  }
  
  expect_equal(result_ls, list())

})

File: python/tests/test_hist.py
# import pytest
import time
import pandas as pd
import yfhist as yfh

# @pytest.mark.skip(reason = "long-running test")

test_symbols = [["AAPL"], ["AAPL", "MSFT"], ["AAPL", "MSFT", "AMZN"]]
test_cols = ["open", "high", "low", "close", "adjclose", "volume"]

def test_that(): # valid 'from_date' and 'interval'

  fields = yfh.data_intervals["field"]

  count = 0
  result_ls = []
  errors_ls = []

  for field in fields:
    
    lookback = yfh.data_intervals.loc[yfh.data_intervals["field"] == field, "lookback"].item()
    
    if pd.isna(lookback):
      lookback = None
    
    if (lookback is None):
      from_date = "2007-01-01"
    else:
      if (field == "1m"):
        
        to_date = pd.Timestamp.now(tz = "UTC")
        from_date = to_date - pd.Timedelta(days = 8)
        
      else:
        
        to_date = None
        from_date = (pd.Timestamp.today().normalize() - pd.Timedelta(days = int(lookback) - 1))

    for symbols in test_symbols:
      
      try:
  
        data = yfh.get_data(symbols, from_date = from_date,
                            to_date = to_date, interval = field)
        
        for col in test_cols:
          response = yfh.get_col(data, col)
        
        # if (response is None):
        #   response = "success"
          
      except:
        response = None
  
      if response is None:
  
        errors_ls.append({
          "symbols": ", ".join(symbols),
          "field": field
        })
  
      count += 1
          
      if (count % 5 == 0):
              
        print("pause one second after five requests")
        time.sleep(1)

  if len(errors_ls) > 0:
    result_ls.extend(errors_ls)

  result_df = pd.DataFrame(result_ls)

  pd.testing.assert_frame_equal(result_df, pd.DataFrame())
